File: 0072-edit-distance/0072-edit-distance.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        # dynamic programming, bottom up
        cache = [[float("inf")]* (len(word2)+1) for _ in range(len(word1)+1)]
        # fill in the bottom row
        for col in range(len(word2)+1): 
            cache[len(word1)][col] = len(word2) - col # base case when word 1 is empty
        # fill in the last column
        for row in range(len(word1)+1): 
            cache[row][len(word2)] = len(word1) - row # base case when word 2 is empty
        for i in range(len(word1) - 1, -1, -1):
            for j in range(len(word2)-1,-1,-1):
                if word1[i] == word2[j]: # char is equal
                    cache[i][j] = cache[i+1][j+1]
                else: # char is not equal
                    cache[i][j] = min(cache[i+1][j],cache[i][j+1],cache[i+1][j+1])+1 # check insert, delete, replace, all there directions in the 2d cache array

        return cache[0][0]

File: 0072-edit-distance/0072-edit-distance.go
func minDistance(word1 string, word2 string) int {
    dp := make([][]int, len(word1) + 1)
    for i := range dp {
        dp[i] = make([]int, len(word2) + 1)
        dp[i][0] = i
    }

    for i := range dp[0] {
        dp[0][i] = i
    }

    for i := 1; i <= len(word1); i++ {
        for j := 1; j <= len(word2); j++ {
            if word1[i-1] == word2[j-1] {
                dp[i][j] = dp[i-1][j-1]
            } else {
                dp[i][j] = min(dp[i][j-1], dp[i-1][j], dp[i-1][j-1]) + 1
            }
        }
    }

    return dp[len(word1)][len(word2)]
}

func min(nums ...int) int {
    ans := nums[0]
    for _, v := range nums {
        if v < ans {
            ans = v
        }
    }

    return ans
}